src/proxy/inbound.rs

```diff
@@ -262,38 +346,90 @@ impl Inbound {
         let hbone_addr = req
             .uri()
             .to_string()
-            .as_str()
             .parse::<SocketAddr>()
-            .map_err(|_| {
-                InboundError(
-                    Error::ConnectAddress(req.uri().to_string()),
-                    StatusCode::BAD_REQUEST,
-                )
-            })?;
+            .map(HboneAddress::SocketAddr)
+            .unwrap_or_else(|_| {
+                let hbone_host = req.uri().host().unwrap();
+                let hbone_port = req.uri().port_u16().unwrap();
+                HboneAddress::SvcHostname(hbone_host.into(), hbone_port)
+            });
 
+        // Get the destination workload information of the destination pods (wds) workload (not destination ztunnel)
         let destination_workload = pi
             .local_workload_information
             .get_workload()
             .await
             // At this point we already fetched the local workload for TLS, so it should be infallible.
             .map_err(InboundError::build(StatusCode::SERVICE_UNAVAILABLE))?;
 
-        // Check the request is allowed
-        Self::validate_destination(&pi.cfg, &pi.state, &conn, &destination_workload, hbone_addr)
+        // Set to the service IP address if the hbone_addr is a hostname
+        // Used for the proxy protocol header. The Address header does not support hostname.
+        let mut upstream_protocol_addr = None;
+        // Check the request is allowed by verifying the destination
+        match hbone_addr {
+            HboneAddress::SocketAddr(_) => Self::validate_destination(
+                &pi.cfg,
+                &pi.state,
+                &conn,
+                &destination_workload,
+                hbone_addr.clone(),
+            )
             .await
-            .map_err(InboundError::build(StatusCode::BAD_REQUEST))?;
+            .map_err(InboundError::build(StatusCode::BAD_REQUEST))?,
+            HboneAddress::SvcHostname(_, _) => {
+                // Get the service address by hostname
+                let addr = Self::find_service_by_hostname(
+                    &pi.state,
+                    &destination_workload,
+                    hbone_addr.clone(),
+                )
+                .await
+                .map_err(InboundError::build(StatusCode::BAD_REQUEST))?;
+
+                // Validate the destination pod is a member of the service
+                match addr {
+                    Address::Service(svc) => {
+                        if !svc.contains_endpoint(&destination_workload) {
+                            // TODO(jaellio): Update error msg/type
+                            return Err(InboundError(
+                                Error::NoResolvedAddresses(destination_workload.to_string()),
+                                StatusCode::BAD_REQUEST,
+                            ));
+                        }
+                        // TODO(jaellio): Intelligently select the VIP
```

Comment:
@keithmattix with regards to https://github.com/istio/ztunnel/pull/1409 and intelligently selecting vip for a Address::Service, is it sufficient to rank vips by network comparison (higher rank if `vip.network` matches `conn.dst_network`) or should we also be comparing IpAddr ranges? Do we also want to implement ranking of Address::Services when finding hostnames (similar to how there would be duplicate workloads on lookup by IP)?

Reply:
Sorry, missed this. Are you still blocked here?

Reply:
Based on [this comment](https://github.com/istio/ztunnel/pull/1420#discussion_r1917454377) we can't rely on the hostname to include the namespace. So right now we'll need to change some of the logic for finding a service by hostname and implement a ranking as suggested [here](https://github.com/istio/ztunnel/pull/1420#discussion_r1917455626)

Reply:
Not blocked, I'll implement the service ranking and vip ranking as suggested above. What just curious if you had any thoughts on the approach.

Reply:
@keithmattix Implemented simple ranking for service (when searching without a namespace) and vip. There might be a simpler way to accomplish the ranking but took inspiration from https://github.com/istio/ztunnel/pull/1409